### PR TITLE
Update Maps SDK 7.x and events 4.x with new location APIs

### LINKED ISSUE #928: Keep sources in memory
When looking into https://github.com/mapbox/mapbox-gl-native/issues/11843. I'm noticing the following in [MapUtils](https://github.com/mapbox/mapbox-navigation-android/blob/d339f623aac301722ac32239276f189641c1bf6c/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/utils/MapUtils.java):

```java
  public static void updateMapSourceFromFeatureCollection(@NonNull MapboxMap mapboxMap,
                                                          @Nullable FeatureCollection collection,
                                                          @NonNull String sourceId) {
    if (collection == null) {
      collection = FeatureCollection.fromFeatures(new Feature[] {});
    }

    GeoJsonSource source = mapboxMap.getSourceAs(sourceId);
    if (source == null) {
      GeoJsonOptions routeGeoJsonOptions = new GeoJsonOptions().withMaxZoom(16);
      GeoJsonSource routeSource = new GeoJsonSource(sourceId, collection, routeGeoJsonOptions);
      mapboxMap.addSource(routeSource);
    } else {
      source.setGeoJson(collection);
    }
  }
```

With every update we are getting the source from MapboxMap, this is an expensive task which requires to go through JNI to C++ an back. Instead of creating new instances with getSourceAs we should cache the value of the source.

### LINKED ISSUE #1121: GPS Not Enabling + Not finding Location
I have since updated our project to the latest version of the SDK, 0.15 and maps sdk to the latest versions. I have now run into the issue that my GPS is not being enabled anymore since the update, which was working fine before the update. I'm not sure if i am maybe missing something now or the docs have excluded something.

`import com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayerPlugin;
 import com.mapbox.android.core.location.LocationEngine;
 import com.mapbox.mapboxsdk.plugins.locationlayer.LocationLayerPlugin;
 import com.mapbox.android.core.location.LocationEngineListener;
 import com.mapbox.android.core.location.LocationEnginePriority;
 import com.mapbox.android.core.location.LocationEngineProvider;
 import com.mapbox.android.core.permissions.PermissionsListener;
 import com.mapbox.android.core.permissions.PermissionsManager;

     public class MainActivity extends AppCompatActivity
        implements NavigationView.OnNavigationItemSelectedListener, LocationEngineListener, PermissionsListener {

    public static LocationEngine locationEngine;
    private LocationLayerPlugin locationPlugin;
    Location lastLocation;
    private PermissionsManager permissionsManager;

        @Override
    protected void onCreate(Bundle savedInstanceState) {
      enableLocationPlugin();
   }

            @SuppressWarnings({"MissingPermission"})
    private void enableLocationPlugin() {

        Log.e("LOC", "ENABLE");

        // Check if permissions are enabled and if not request
        if (PermissionsManager.areLocationPermissionsGranted(this)) {
            // Create an instance of LOST location engine

            initializeLocationEngine();

        } else {

            permissionsManager = new PermissionsManager(this);
            permissionsManager.requestLocationPermissions(this);
        }
    }

    @SuppressLint("MissingPermission")
    private void initializeLocationEngine() {
        LocationEngineProvider locationEngineProvider = new LocationEngineProvider(this);
        locationEngine = locationEngineProvider.obtainBestLocationEngineAvailable();
        locationEngine.setPriority(LocationEnginePriority.HIGH_ACCURACY);

        Location lastLocation = locationEngine.getLastLocation();
        if (lastLocation != null && lastLocation.getLongitude() != 0 && lastLocation.getLatitude() != 0) {
            originLocation = lastLocation;
            setCameraPosition(lastLocation);
        } else {
            locationEngine.addLocationEngineListener(this);
            locationEngine.activate();
            Log.e("LOCATION", "NULL LOCATION");
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        permissionsManager.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    public void onExplanationNeeded(List<String> permissionsToExplain) {

    }

    @Override
    public void onPermissionResult(boolean granted) {
        if (granted) {
            enableLocationPlugin();
        } else {
            finish();
        }
    }

    @Override
    public void onConnected() {

        if (locationEngine != null) {
            locationEngine.requestLocationUpdates();
                lastLocation = locationEngine.getLastLocation()
    }

`

I noticed in the onConnected() method it is now saying it now requires permission which may be rejected by the user on both the requestLocationUpdates and getLastLocation(). This was never underlined previously.

My lastLocation is always coming back as null since the update and the gps doesn't seem to get enabled. What have i missed?

I added only the necessary code as an example.

### PR DESCRIPTION
**Pull request by danesfeder**
Closes #1121 Closes #928

This PR brings in the new `Location` APIs from events `4.x`.  This is a decent refactor as we use location services a lot and in almost every activity of the test app. 

TODO:
- [x] Fix-up remaining test activities 
- [x] Add / update tests
- [x] Bring in correct Maps SDK dependency

### CODE REVIEW

**Comment by danesfeder** [comment]
@andrlee I'm cutting this a bit prematurely to get eyes on it as soon as possible.  Please try to ignore the mess.

**Comment by LukasPaczos** [comment]
Just wanted to flag https://github.com/mapbox/mapbox-events-android/issues/312 as a potential issue that will need a workaround until fixed upstream. Workaround tracked for the Maps SDK in https://github.com/mapbox/mapbox-gl-native/issues/13587.

**Comment by Guardiola31337** [comment]
While testing this, I run into some issues.

`TODO` here:

- [ ] Leaks originated by new location engine
  - [ ] `NavigationLauncherActivity`
  - [ ] `ComponentNavigationActivity`
- [ ] Revisit `ReplayRouteLocationEngine` implementation
  - [ ] `EndNavigationActivity`

```
java.lang.Exception: location can't be null
            at com.mapbox.services.android.navigation.v5.location.replay.ReplayRouteLocationEngine.getLastLocation(ReplayRouteLocationEngine.java:102)
```

```
java.lang.Exception: No route found to replay.
            at com.mapbox.services.android.navigation.v5.location.replay.ReplayRouteLocationEngine.requestLocationUpdates(ReplayRouteLocationEngine.java:115)
```

```
java.lang.Exception: location can't be null
            at com.mapbox.services.android.navigation.v5.location.replay.ReplayRouteLocationEngine.getLastLocation(ReplayRouteLocationEngine.java:102)
```

  - [ ] `RerouteActivity` - doesn’t allow re-routes
  - [ ] `WaypointNavigationActivity` - doesn’t navigate to next destination
- [ ] `MockNavigationActivity` doesn’t fetch routes (`calculateRoute` is commented out) - code needs to be adapted to new location lib changes `locationEngine.getLastLocation()` is not available anymore)

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/LocationUpdater.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  }
+
+  void removeLocationUpdates() {
+    locationEngine.removeLocationUpdates(callback);
+  }
+
+  @SuppressLint("MissingPermission")
+  private void requestInitialLocationUpdates(LocationEngine locationEngine, LocationEngineRequest request) {
+    locationEngine.getLastLocation(callback);
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/LocationUpdater.java]: What happens in the situation in which the last location is really far from where you actually are? Could this end up in a "undesired" flying effect? I'm wondering if we should / could remove it 🤔

**Comment by Guardiola31337** [comment]
Run into the following leak while testing `EndNavigationActivity` 👀 

leak_end_navigation_activity

Not sure if caused by the changes introduced here. If not, we should cut a separate ticket to investigate further.

**Comment by Guardiola31337** [comment]
I've also noticed that the puck flies to _Null Island_ (and then comes back) when reaching the first waypoint in `EndNavigationActivity` 👀

null_island_end_navigation_activity

**Comment by danesfeder** [comment]
@kevinkreiser is looking into https://github.com/mapbox/mapbox-navigation-android/pull/1615#issuecomment-452682417 from a `Navigator` standpoint.

**Comment by danesfeder** [comment]
Per discussion with @Guardiola31337, we have one remaining issue regarding allowing users to zoom in too closely to the map.  He's going to pick that up and then bring this PR home 🎉 

The rest of the issues we found while testing seem unrelated to the Maps update and we will ticket as necessary.

**Comment by Guardiola31337** [comment]
> we have one remaining issue regarding allowing users to zoom in too closely to the map

This is now fixed 🎉 

Noting that the zoom levels in the `LocationComponent` aren't restricted (by default) anymore. This was disabled in https://github.com/mapbox/mapbox-gl-native/issues/13097 / https://github.com/mapbox/mapbox-gl-native/pull/13425

Now you should tune the the max / min zoom level of the map yourself with `MapboxMapOptions#maxZoomPreference` / `MapboxMapOptions#minZoomPreference` or `MapboxMap#setMaxZoomPreference` / `MapboxMap#setMinZoomPreference` Thanks @LukasPaczos for clarifying 🙏

**Comment by johnbulalo** [comment]
image
it keeps showing this on android studio 3.5

**Pull request merged**